builtin/common/serialize.lua

```diff
@@ -1,205 +1,232 @@
-						..dump_or_ref_val(v)
+			-- Now write map keys ([key] = value)
+			for k, v in next, value do
+				-- Check whether this is a non-list key (hash key)
+				if type(k) ~= "number" or k % 1 ~= 0 or k < 1 or k > len then
+					if use_short_key(k) then
+						write(k)
+					else
+						write("[")
+						dump(k)
+						write("]")
+					end
+					write("=")
+					dump(v)
+					write(";")
 				end
 			end
-			return "{"..table.concat(vals, ", ").."}"
-		else
-			error("Can't serialize data of type "..tp)
+			write("}")
+			return
 		end
 	end
-
-	local function dump_nest_points()
-		for parent, vals in pairs(nest_points) do
-			for k, v in pairs(vals) do
-				local_defs[#local_defs + 1] = dump_or_ref_val(parent)
-					.."["..dump_or_ref_val(k).."] = "
-					..dump_or_ref_val(v)
+	for table, ref in pairs(to_fill) do
+		for k, v in pairs(table) do
+			write(ref)
+			if use_short_key(k) then
+				write(".")
+				write(k)
+			else
+				write("[")
+				dump(k)
+				write("]")
 			end
+			write("=")
+			dump(v)
+			write(";")
 		end
 	end
+	write("return ")
+	dump(value)
+end
 
-	mark_multiple_occurences(x)
-	local top_level = dump_or_ref_val(x)
-	dump_nest_points()
-
-	if next(local_defs) then
-		return "local _ = {}\n"
-			..table.concat(local_defs, "\n")
-			.."\nreturn "..top_level
-	else
-		return "return "..top_level
-	end
+function core.serialize(value)
+	local rope = {}
+	serialize(value, function(text)
+		table_insert(rope, text)
+	end)
+	return table_concat(rope)
 end
 
--- Deserialization
+local function dummy_func() end
 
-local function safe_loadstring(...)
-	local func, err = loadstring(...)
-	if func then
-		setfenv(func, {})
-		return func
+local function deserialize(func, safe)
+	-- math.huge is serialized to inf, 0/0 is serialized to -nan
+	local env = {inf = math_huge, nan = 0/0}
+	if safe then
+		env.loadstring = dummy_func
+	else
+		env.loadstring = function(str, ...)
+			if str:byte(1) == 0x1B then
+				return nil, "Bytecode prohibited"
+			end
+			local func, err = loadstring(str, ...)
+			if func then
+				setfenv(func, env)
+				return func
+			end
+			return nil, err
+		end
 	end
-	return nil, err
+	setfenv(func, env)
+	local success, value_or_err = pcall(func)
+	if success then
+		return value_or_err
+	end
+	return nil, value_or_err
 end
 
-local function dummy_func() end
-
 function core.deserialize(str, safe)
-	if type(str) ~= "string" then
-		return nil, "Cannot deserialize type '"..type(str)
-			.."'. Argument must be a string."
-	end
 	if str:byte(1) == 0x1B then
 		return nil, "Bytecode prohibited"
 	end
-	local f, err = loadstring(str)
-	if not f then return nil, err end
-
-	-- The environment is recreated every time so deseralized code cannot
-	-- pollute it with permanent references.
-	setfenv(f, {loadstring = safe and dummy_func or safe_loadstring})
-
-	local good, data = pcall(f)
-	if good then
-		return data
-	else
-		return nil, data
-	end
+	return deserialize(assert(loadstring(str)), safe)
```

Comment:
why not inline this?

Reply:
Inline what? This comment let me to restore backwards compat regarding serialized data that isn't valid Lua BTW.